src/appleseed.bench/utility/backendapi.cpp

```diff
@@ -0,0 +1,239 @@
+
+    // Don't just assert, abort if the key is too short.
+    if (encryption_key.size() < api_key.size())
+        throw std::runtime_error("Encryption key must be at least as long as backend API key");
+
+    QByteArray encrypted_api_key = api_key;
+    for (int i = 0, e = encrypted_api_key.size(); i < e; ++i)
+        encrypted_api_key[i] = encrypted_api_key[i] ^ encryption_key[i];
+
+    return encrypted_api_key.toBase64();
+}
+
+#endif
+
+QString get_decrypted_api_key()
+{
+    const QByteArray encrypted_api_key = QByteArray::fromBase64(APPLESEED_TO_STRING_EVAL(APPLESEED_BENCH_BACKEND_API_KEY));
+    const QByteArray encryption_key = get_encryption_key();
+
+    assert(encryption_key.size() >= encrypted_api_key.size());
+
+    QByteArray decrypted_api_key = encrypted_api_key;
+    for (int i = 0, e = decrypted_api_key.size(); i < e; ++i)
+        decrypted_api_key[i] = decrypted_api_key[i] ^ encryption_key[i];
+
+    return decrypted_api_key.toBase64();
+}
+
+QNetworkRequest make_api_request(const QString& route)
+{
+    assert(!route.isEmpty() && route[0] == '/');
+
+    const QUrl url(
+        QString("https://appleseed-bench-backend.herokuapp.com%1?apiKey=%2")
+            .arg(route)
+            .arg(get_decrypted_api_key()));
+
+    return QNetworkRequest(url);
+}
+
+QString filter_api_key(const QString& s)
+{
+    QString result = s;
+    result.replace(get_decrypted_api_key(), "XXX");
+    return result;
+}
+
+namespace
+{
+    QString get_string(const QJsonObject& json_obj, const QString& key)
+    {
+        const QJsonValue json_value = json_obj[key];
+        if (!json_value.isString())
+            throw ExceptionJsonValue();
+        return json_value.toString();
+    }
+
+    int get_int(const QJsonObject& json_obj, const QString& key)
+    {
+        const QJsonValue json_value = json_obj[key];
+        bool ok = false;
+        const int value = json_value.toVariant().toInt(&ok);
+        if (!ok)
+            throw ExceptionJsonValue();
+        return value;
+    }
+}
+
+BenchmarkResult BenchmarkResult::from_json_object(const QJsonObject& json_obj)
+{
+    BenchmarkResult result;
+    result.m_id = get_string(json_obj, "_id");
+    result.m_submission_datetime_utc = QDateTime::fromString(get_string(json_obj, "submissionDateTimeUtc"));
+    result.m_benchmark_version = get_int(json_obj, "benchmarkVersion");
+    result.m_benchmark_scene_id = get_string(json_obj, "benchmarkSceneId");
+    result.m_cpu_model = get_string(json_obj, "cpuModel");
+    result.m_enabled_cpu_core_count = get_int(json_obj, "cpuCoreCount");
+    result.m_cpu_thread_count = get_int(json_obj, "cpuThreadCount");
+    result.m_render_thread_count = get_int(json_obj, "renderThreadCount");
+    result.m_render_time = get_int(json_obj, "renderTime");
+    return result;
+}
+
+QJsonObject BenchmarkResult::to_json_object() const
+{
+    QJsonObject json_obj;
+    // _id is ignored, it will be set automatically by MongoDB.
```

Comment:
If needed, automatic adding of `_id` can be avoided.

Reply:
We actually need `_id`, it's used to identify the result that the user submitted when displaying the Rankings window (to highlight the user's result in orange).